landoapi/hgexports.py

```diff
@@ -393,6 +393,147 @@ def get_timestamp(self) -> str:
+             bug |
+             b= |
+             # a sequence of 5+ numbers preceded by whitespace
+             (?=\b\#?\d{5,}) |
+             # numbers at the very beginning
+             ^(?=\d)
+           )
+           (?:\s*\#?)(\d+)(?=\b)
+         )""",
+    re.I | re.X,
+)
+
+
+def is_backout(commit_desc: str) -> bool:
+    """Returns True if commit description indicates the changeset is a backout.
+
+    Backout commits should always result in is_backout() returning True,
+    and parse_backouts() not returning None.  Malformed backouts may return
+    True here and None from parse_backouts().
+    """
+    return BACKOUT_KEYWORD_RE.match(commit_desc) is not None
+
+
+def parse_bugs(commit_desc: str, conservative: bool = False) -> list[int]:
+    m = RE_SOURCE_REPO.search(commit_desc)
+    if m:
+        source_repo = m.group(1)
+
+        if source_repo.startswith("https://github.com/"):
+            conservative = True
+
+    if commit_desc.startswith("Bumping gaia.json"):
+        conservative = True
+
+    bugzilla_re = BUG_CONSERVATIVE_RE if conservative else BUG_RE
+
+    bugs_with_duplicates = [int(m[1]) for m in bugzilla_re.findall(commit_desc)]
+    bugs_seen = set()
+    bugs_seen_add = bugs_seen.add
+    bugs = [x for x in bugs_with_duplicates if not (x in bugs_seen or bugs_seen_add(x))]
+    return [bug for bug in bugs if bug < 100000000]
+
+
+def parse_backouts(
+    commit_desc: str, strict: bool = False
+) -> Optional[tuple[list[str], list[int]]]:
+    """Look for backout annotations in a string.
+
+    Returns a 2-tuple of (nodes, bugs) where each entry is an iterable of
+    changeset identifiers and bug numbers that were backed out, respectively.
+    Or return None if no backout info is available.
+
+    Setting `strict` to True will enable stricter validation of the commit
+    description (eg. ensuring N commits are provided when given N commits are
+    being backed out).
+    """
+    if not is_backout(commit_desc):
+        return None
+
+    lines = commit_desc.splitlines()
+    first_line = lines[0]
+
+    # Single backout.
+    m = BACKOUT_SINGLE_RE.match(first_line)
+    if m:
+        return [m.group("node")], parse_bugs(first_line)
+
+    # Multiple backouts, with nodes listed in commit description.
+    m = BACKOUT_MULTI_SPLIT_RE.match(first_line)
+    if m:
+        expected = int(m.group("count"))
+        nodes = []
+        for line in lines[1:]:
+            single_m = BACKOUT_SINGLE_RE.match(line)
+            if single_m:
+                nodes.append(single_m.group("node"))
+        if strict:
+            # The correct number of nodes must be specified.
+            if expected != len(nodes):
+                return None
+        return nodes, parse_bugs(commit_desc)
+
+    # Multiple backouts, with nodes listed on the first line
+    m = BACKOUT_MULTI_ONELINE_RE.match(first_line)
+    if m:
+        return SHORT_NODE_RE.findall(m.group("nodes")), parse_bugs(first_line)
+
+    return None
```

Comment:
There is a mix of `return` and `return None` in this PR. Would be good to pick one and stick with it, at least for this PR.